src/rust/src/lib.rs

```diff
@@ -2,9 +2,93 @@
 // 2.0, and the BSD License. See the LICENSE file in the root of this repository
 // for complete details.
 
+use std::convert::TryInto;
+
+/// Returns the value of the input with the most-significant-bit copied to all
+/// of the bits.
+fn duplicate_msb_to_all(a: u8) -> u8 {
+    0u8.wrapping_sub(a >> 7)
+}
+
+/// This returns 0xFF if a < b else 0x00, but does so in a constant time
+/// fashion.
+fn constant_time_lt(a: u8, b: u8) -> u8 {
+    // Derived from:
+    // https://github.com/openssl/openssl/blob/OpenSSL_1_1_1i/include/internal/constant_time.h#L120
+    duplicate_msb_to_all(a ^ ((a ^ b) | (a.wrapping_sub(b) ^ b)))
+}
+
+#[pyo3::prelude::pyfunction]
+fn check_pkcs7_padding(data: &[u8]) -> bool {
+    assert!(!data.is_empty());
+
+    let mut mismatch = 0;
+    let pad_size = data[data.len() - 1];
+    let len: u8 = data.len().try_into().expect("data too long");
+    for (i, b) in (0..len).zip(data.iter().rev()) {
+        let mask = constant_time_lt(i, pad_size);
+        mismatch |= mask & (pad_size ^ b);
+    }
+
+    // Check to make sure the pad_size was within the valid range.
+    mismatch |= !constant_time_lt(0, pad_size);
+    mismatch |= constant_time_lt(data.len() as u8, pad_size);
```

Comment:
We have `len` now, no need for another (theoretically) lossy cast.
```suggestion
    mismatch |= constant_time_lt(len, pad_size);
```